File: ParserGenerator.scala
import java.io.{File, PrintWriter}

import scala.collection.mutable


/**
  * this class generates parsers for Kotlin data classes,
  *
  * it is null safe, all not null type (T, not T?) has a default value, which is not null
  *
  * for T?, it is null
  * for numbers, it is 0 of the type
  * for string, it is ""
  * for boolean, it is false
  * for data objects, it is the data object will all fields of the default value
  *
  * TODO if these defaults is not what we have in mind, we need to created a withDefault method
  *
  * thus, it is safe to declare a type as not null, even when the json string is null for the type
  *
  * when this happens, it will print a warning
  */



val ExtraCodeStart = "/* EXTRA CODE START */"
val ExtraCodeMark = "/* EXTRA CODE MARK */"
val ExtraCodeEnd = "/* EXTRA CODE END */"

val ParseFinishStart = "/* PARSE FINISH START */"
val ParseFinishMark = "/* PARSE FINISH MARK */"
val ParseFinishEnd = "/* PARSE FINISH END */"

sealed abstract class Type(val name: String) {

  def default(): String
  def parserObject(): String
  def parse(): String

  def ? = NullableType(this)
  def nullName = if (name.endsWith("?")) name else  name + "?"
}

case object IntType extends Type("Int") {
  override def default() = "0"
  override def parse() = "jp.valueAsInt"
  override def parserObject() = "JsonAdapter.intAdapter"
}
case object LongType extends Type("Long") {
  override def default() = "0L"
  override def parse() = "jp.valueAsLong"
  override def parserObject() = "JsonAdapter.longAdapter"
}
case object DoubleType extends Type("Double") {
  override def default() = "0.0"
  override def parse() = "jp.valueAsDouble"
  override def parserObject() = "JsonAdapter.doubleAdapter"
}
case object BooleanType extends Type("Boolean") {
  override def default() = "false"
  override def parse() = "jp.valueAsBoolean"
  override def parserObject() = "JsonAdapter.booleanAdapter"
}
case object StringType extends Type("String") {
  override def default() = "\"\""
  override def parse() = "jp.getValueAsString(\"\")"
  override def parserObject() = "JsonAdapter.stringAdapter"
}
case class Field(name: String, ty: Type, jsonName: Option[String], isPrivate: Boolean = false) {

  val jn= jsonName.getOrElse(name)
  def parse() = s""""$jn" -> { $name = ${ty.parse()} }"""

  def declareTemp() = s"var $name: ${ty.name} = ${ty.default()}"

  def tempToField() = name // if (ty.isInstanceOf[NullableType]) name else name + "!!"

  def declare() = s"        ${(if (isPrivate) "private " else "@JvmField ")}val $name: ${ty.name}"

  def serialize() = ty match {
    case IntType | LongType | DoubleType => s"""jg.writeNumberField("$jn", t.$name)"""
    case StringType => s"""jg.writeStringField("$jn", t.$name)"""
    case BooleanType => s"""jg.writeBooleanField("$jn", t.$name)"""
    case NullableType(t) => s"""if (t.$name != null) { jg.writeFieldName("$jn"); ${t.parserObject()}.serialize(t.$name, jg, true) }"""
    case _ => s"""jg.writeFieldName("$jn"); ${ty.parserObject()}.serialize(t.$name, jg, true)"""
  }
}

def indent(in: String, statement: String) = {
  statement.split("\n").filter(!_.trim.isEmpty).map(in + _).mkString("\n")
}

case class ListType(t: Type) extends Type(s"List<${t.name}>") {
  override def default() = "emptyList()"
  override def parserObject() = t.parserObject() +".arrayAdapter()"
  override def parse() = parserObject() + ".parse(jp)"
}

case class JvmType(override val name: String, defaultStr: String) extends Type(name) {
  override def default() = defaultStr
  override def parserObject() = throw new Exception("Not allowed")
  override def parse() = throw new Exception("Not allowed")
}

case class HandmadeObjectType(override val name: String) extends Type(name) {
  override def default() = s"${name}.Companion.empty"
  override def parserObject() = name + ".Companion"
  override def parse() = parserObject() + ".parse(jp)"
}

case class ObjectType(override val name: String, fields: Seq[Field]) extends Type(name) {
  override def default() = s"${name}.Companion.empty"
  def codgen() =
    s"""
       |
       |data class $name(
       |${fields.map(_.declare()).mkString(",\n")})  : Serializable  {
       |  companion object : ObjectJsonAdapter<$name>() {
       |    override val empty: $name = $name(${fields.map(_.ty.default()).mkString(", ")})
       |    override fun parse(jp: JsonParser): $name {
       |      // Our code ensures all parse method will not throw error, if null encountered
       |      if (jp.currentToken != JsonToken.START_OBJECT) {
       |        jp.skipChildren()
       |        return empty
       |      }
       |${indent("      ", fields.map(_.declareTemp()).mkString("\n"))}
       |      while (jp.nextToken() != JsonToken.END_OBJECT) {
       |        val fieldName = jp.currentName
       |        jp.nextToken()
       |        when(fieldName) {
       |${indent("          ", fields.map(_.parse()).mkString("\n"))}
       |          else -> { logUnknownField(fieldName, ${name}.Companion) }
       |        }
       |        jp.skipChildren()
       |      }
       |      return $name(${fields.map(_.tempToField()).mkString(", ")})
       |    }
       |
       |    override fun serializeFields(t: $name, jg: JsonGenerator) {
       |${indent("      ", fields.map(_.serialize()).mkString("\n"))}
       |    }
       |  }
       |$ExtraCodeMark
       |}
     """.stripMargin

  override def parserObject() = name + ".Companion"

  override def parse() = parserObject() + ".parse(jp)"
}

case class NullableType(t: Type) extends Type(t.name + "?") {
  override def default() = "null"
  override def parserObject() = t.parserObject() +".nullAdapter()"
  override def parse() = parserObject() + ".parse(jp)"
}
case class EnumType(override val name: String, vs: Seq[String], defaultPos: Int) extends Type(name) {
  override def default() = name + "." + vs(defaultPos)
  def codgen() = s"""
  |enum class $name {
  |  ${vs.mkString(", ")};
  |  companion object : EnumJsonAdapter<$name>(arrayOf(${vs.map(a => name + "." + a).mkString(", ")}), $name.${vs(defaultPos)})
  |}
  """.stripMargin

  override def parserObject() = s"${name}.Companion"
  override def parse() = parserObject() + ".parse(jp)"
}

case class ConvertedType(json: Type, jvm: Type, converter: String) extends Type(jvm.name) {
  override def default() = jvm.default()

  override def parserObject() = converter

  override def parse() = parserObject() + ".parse(jp)"
}


val int = IntType
val long = LongType
val double = DoubleType
val boolean = BooleanType
val string = StringType

def list(t: Type) = ListType(t)


def f(name: String, ty: Type) = Field(name, ty, None)
def f(name: String,  json: String, ty: Type) = Field(name, ty, Some(json))


lazy val DefaultImports =
  s"""
import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonToken
import android.util.Log
import android.net.Uri
import java.io.*
import java.util.*
import android.text.TextUtils
import android.util.Log
import org.snailya.kotlinparsergenerator.*
import java.io.IOException
import java.text.SimpleDateFormat
""".stripMargin

class Spec(packageName: String, root: File, imports: Seq[String]) {


  lazy val dest = new File(root + "/" + packageName.replace('.', '/'))

  def writeClass(className: String, cont: String): Unit = {
    println("writing class " + className)
    var content = cont
    val des = new File(dest, className + ".kt")
    des.mkdirs()
    var imps = Vector.empty[String]
    if (des.isFile) {
      val lines = io.Source.fromFile(des).getLines().toSeq
      var extra = ""
      var started = false
      for (line <- lines) {
        if (line.startsWith("import ") && !imps.contains(line.trim)) {
          imps = imps :+ line
        }
        if (line.contains(ExtraCodeStart)) {
          started = true
        } else if (line.contains(ExtraCodeEnd)) {
          started = false
        } else if (started) {
          extra = extra + line + "\n"
        }
      }
      if (!started) {
        extra = ExtraCodeStart + "\n" + extra + ExtraCodeEnd
        content = content.replace(ExtraCodeMark, extra)
      }
      var parseComplete = ""
      started = false
      for (line <- lines) {
        if (line.contains(ParseFinishStart)) {
          started = true
        } else if (line.contains(ParseFinishEnd)) {
          started = false
        } else if (started) {
          parseComplete = parseComplete + line + "\n"
        }
      }
      if (!started) {
        parseComplete = ParseFinishStart + "\n" + parseComplete + ParseFinishEnd
        content = content.replace(ParseFinishMark, parseComplete)
      }
    }
    content = "package " + packageName + ";\n\n" + (DefaultImports + (imps ++ imports).mkString("\n")).split("\n").distinct.mkString("\n") + "\n" + content
    des.delete()
    des.createNewFile()
    val writer = new PrintWriter(des)
    writer.write(content)
    writer.close()
  }

  def writeFile(des: File, cont: String): Unit = {
    des.mkdirs()
    des.delete()
    des.createNewFile()
    val writer = new PrintWriter(des)
    writer.write(cont)
    writer.close()
  }


  val os = mutable.ArrayBuffer.empty[ObjectType]
  val es = mutable.ArrayBuffer.empty[EnumType]
  def o(s: String, fs: Field*) = {
    val o = ObjectType(s, fs)
    os.append(o)
    o
  }

  def e(s: String, de: Int, fs: String*) = {
    val o = EnumType(s, fs, de)
    es.append(o)
    o
  }

  def codegen(): Unit = {
    es.foreach(e => writeClass(e.name, e.codgen()))
    os.foreach(o => writeClass(o.name, o.codgen()))
  }
}


object BaseSpec extends Spec("org.snailya.kotlinparsergenerator.data", new File("app/src/main/java"), Seq.empty) {

  lazy val stringToUri = ConvertedType(string, JvmType("Uri", "defaultUri").?, "stringToUri")

  e("SomeEnum", 0, "enum1", "enum2")

  o("Location",
    f("latitude", double),
    f("longitude", double),
    f("coordinateType", HandmadeObjectType("CoordinateType").?),
    f("countryCode", string.?),
    f("countryName", string.?),
    f("cityName", string.?),
    f("someArray", list(string)),
    f("someUrl", stringToUri),
    f("cached", boolean)
  )
}

BaseSpec.codegen()

File: generator/src/main/java/org/snailya/parsergenerator/ParserGenerator.kt
package org.snailya.parsergenerator

import java.io.File
import java.io.PrintWriter
import java.util.*
import java.util.Collections.emptyList

/**
  * this class generates parsers for Kotlin data classes,
  *
  * it is null safe, all not null type (T, not T?) has a default value, which is not null
  *
  * for T?, it is null
  * for numbers, it is 0 of the type
  * for string, it is ""
  * for boolean, it is false
  * for data objects, it is the data object will all fields of the default value
  *
  * TODO if these defaults is not what we have in mind, we need to created a withDefault method
  *
  * thus, it is safe to declare a type as not null, even when the json string is null for the type
  *
  * when this happens, it will print a warning
  */


val ExtraCodeStart = "/* EXTRA CODE START */"
val ExtraCodeMark = "/* EXTRA CODE MARK */"
val ExtraCodeEnd = "/* EXTRA CODE END */"

val ParseFinishStart = "/* PARSE FINISH START */"
val ParseFinishMark = "/* PARSE FINISH MARK */"
val ParseFinishEnd = "/* PARSE FINISH END */"

fun indent(i: String, statement: String): String =
  statement.split("\n").filter{ !it.trim().isEmpty() }.map {i + it }.joinToString("\n")



abstract class Type(open val name: String)  {

  abstract fun default(): String
  abstract fun parserObject(): String
  abstract fun parse(): String

  val n get() = NullableType(this)
  val nullName get() = if (name.endsWith("?")) name else  name + "?"
}

object IntType : Type("Int") {
  override fun default() = "0"
  override fun parse() = "jp.valueAsInt"
  override fun parserObject() = "JsonAdapter.intAdapter"
}
object LongType : Type("Long") {
  override fun default() = "0L"
  override fun parse() = "jp.valueAsLong"
  override fun parserObject() = "JsonAdapter.longAdapter"
}
object DoubleType : Type("Double") {
  override fun default() = "0.0"
  override fun parse() = "jp.valueAsDouble"
  override fun parserObject() = "JsonAdapter.doubleAdapter"
}
object BooleanType : Type("Boolean") {
  override fun default() = "false"
  override fun parse() = "jp.valueAsBoolean"
  override fun parserObject() = "JsonAdapter.booleanAdapter"
}
object StringType : Type("String") {
  override fun default() = "\"\""
  override fun parse() = "jp.getValueAsString(\"\")"
  override fun parserObject() = "JsonAdapter.stringAdapter"
}


data class NullableType(val t: Type) : Type(t.name + "?") {
  override fun default() = "null"
  override fun parserObject() = t.parserObject() +".nullAdapter()"
  override fun parse() = parserObject() + ".parse(jp)"
}

data class ListType(val t: Type) : Type("List<${t.name}>") {
  override fun default() = "emptyList()"
  override fun parserObject() = t.parserObject() +".arrayAdapter()"
  override fun parse() = parserObject() + ".parse(jp)"
}

data class JvmType(override val name: String, val defaultStr: String) : Type(name) {
  override fun default() = defaultStr
  override fun parserObject() = throw Exception("Not allowed")
  override fun parse() = throw Exception("Not allowed")
}

data class HandmadeObjectType(override val name: String) : Type(name) {
  override fun default() = "$name.Companion.empty"
  override fun parserObject() = name + ".Companion"
  override fun parse() = parserObject() + ".parse(jp)"
}

data class Field(val name: String, val ty: Type, val jsonName: String?, val isPrivate: Boolean = false, val deprecated: String = "", var isOverride: Boolean = false) {

  val jn= jsonName ?: name
  fun parse() = """"$jn" -> { $name = ${ty.parse()} }"""

  fun declareTemp() ="var $name: ${ty.name} = ${ty.default()}"

  fun tempToField() = name // if (ty.isInstanceOf[NullableType]) name else name + "!!"

  fun declare() = "        ${ (if (!deprecated.isEmpty()) "@Deprecated(\"$deprecated\") " else "")}${(if (isPrivate) "private " else "@JvmField ")}${(if (isOverride) "override " else "")}val $name: ${ty.name}"

  fun serialize() =
    if (ty == IntType || ty == DoubleType || ty == LongType) {
      """jg.writeNumberField("$jn", t.$name)"""
    } else if (ty == StringType) {
      """jg.writeStringField("$jn", t.$name)"""
    } else if (ty == BooleanType) {
      """jg.writeBooleanField("$jn", t.$name)"""
    } else if (ty is NullableType) {
      """if (t.$name != null) { jg.writeFieldName("$jn"); ${ty.t.parserObject()}.serialize(t.$name, jg, true) }"""
    } else {
      """jg.writeFieldName("$jn"); ${ty.parserObject()}.serialize(t.$name, jg, true)"""
    }
}



data class ObjectType(override val name: String, private val fields: List<Field>, var implements: String? = null) : Type(name) {
  override fun default() = "$name.Companion.empty"

  private val extraImplements get() = if (implements != null) ("," + implements) else ""

  fun codgen() =
    """
       |
       |data class $name(
       |${fields.map{it.declare()}.joinToString(",\n")})  : Serializable $extraImplements {
       |
       |$ExtraCodeMark
       |
       |
       |  // BELLOW IS GENERATED CODE!!! DON'T CHANGE!!!
       |
       |  companion object : ObjectJsonAdapter<$name>() {
       |    override val empty: $name = $name(${fields.map { it.ty.default() }.joinToString(", ")})
       |    override fun parse(jp: JsonParser): $name {
       |      // Our code ensures all parse method will not throw error, if null encountered
       |      if (jp.currentToken != JsonToken.START_OBJECT) {
       |        jp.skipChildren()
       |        return empty
       |      }
       |${indent("      ", fields.map { it.declareTemp() }.joinToString("\n"))}
       |      while (jp.nextToken() != JsonToken.END_OBJECT) {
       |        val fieldName = jp.currentName
       |        jp.nextToken()
       |        when(fieldName) {
       |${indent("          ", fields.map { it.parse()}.joinToString("\n"))}
       |          else -> { JsonAdapterGlobalConfig.logUnknownField(fieldName, ${name}.Companion) }
       |        }
       |        jp.skipChildren()
       |      }
       |      return $name(${fields.map { it.tempToField()}.joinToString(", ")})
       |    }
       |
       |    override fun serializeFields(t: $name, jg: JsonGenerator) {
       |${indent("      ", fields.map { it.serialize()}.joinToString("\n"))}
       |    }
       |  }
       |}
     """.trimMargin()

  override fun parserObject() = name + ".Companion"

  override fun parse() = parserObject() + ".parse(jp)"
}

data class IntEnumItem(val name: String, val value: Int)


data class IntEnumType(override val name: String, val vs: List<IntEnumItem>, val defaultPos: Int) : Type(name) {
  override fun default() = name + "." + vs[defaultPos].name
  fun codgen() = """
  |enum class $name(override val value: Int) : IntEnum {
  |  ${vs.joinToString(", ") { "${it.name}(${it.value})"}};
  |  companion object : IntEnumJsonAdapter<$name>(arrayOf(${vs.map{a -> name + "." + a.name}.joinToString(", ")}), $name.${vs[defaultPos].name})
  |  override fun toString(): String = Companion.serialize(this)
  |}
  """.trimMargin()

  override fun parserObject() = "${name}.Companion"
  override fun parse() = parserObject() + ".parse(jp)"
}

data class StringEnumItem(val name: String, val value: String)


data class StringEnumType(override val name: String, val vs: List<StringEnumItem>, val defaultPos: Int) : Type(name) {
  override fun default() = name + "." + vs[defaultPos].name
  fun codgen() = """
  |enum class $name(override val value: String) : StringEnum {
  |  ${vs.joinToString(", ") { "${it.name}(\"${it.value}\")"}};
  |  companion object : StringEnumJsonAdapter<$name>(arrayOf(${vs.map{a -> name + "." + a.name}.joinToString(", ")}), $name.${vs[defaultPos].name})
  |  override fun toString(): String = Companion.serialize(this)
  |}
  """.trimMargin()

  override fun parserObject() = "${name}.Companion"
  override fun parse() = parserObject() + ".parse(jp)"
}




data class ConvertedType(val json: Type, val jvm: Type, val converter: String) : Type(jvm.name) {
  override fun default() = jvm.default()

  override fun parserObject() = converter

  override fun parse() = parserObject() + ".parse(jp)"
}




open class Spec(val packageName: String, val root: File, val imports: List<String>) {


  init {
    println("root is ${root.absolutePath}")
  }
  val int = IntType
  val long = LongType
  val double = DoubleType
  val boolean = BooleanType
  val string = StringType

  fun list(t: Type) = ListType(t)


  fun f(name: String, ty: Type, deprecated: String = "") = Field(name, ty, null, deprecated = deprecated)
  fun f(name: String,  json: String, ty: Type, deprecated: String = "") = Field(name, ty, json, deprecated = deprecated)

  open val defaultImports =
      """
import com.fasterxml.jackson.core.*
import java.io.*
import java.util.*
import org.snailya.kotlinparsergenerator.*
""".trimMargin()

  val dest = File(root.toString() + "/" + packageName.replace('.', '/'))

  fun writeClass(className: String, cont: String) {
    println("writing class " + className)
    var content = cont
    val des = File(dest, className + ".kt")
    des.mkdirs()
    var imps = ArrayList<String>()
    if (des.isFile) {
      val lines: List<String> = des.readLines()
      var extra = ""
      var started = false
      for (line in lines) {
        if (line.startsWith("import ") && !imps.contains(line.trim())) {
          imps.add(line)
        }
        if (line.contains(ExtraCodeStart)) {
          started = true
        } else if (line.contains(ExtraCodeEnd)) {
          started = false
        } else if (started) {
          extra = extra + line + "\n"
        }
      }
      if (!started) {
        extra = ExtraCodeStart + "\n" + extra + ExtraCodeEnd
        content = content.replace(ExtraCodeMark, extra)
      }
      var parseComplete = ""
      started = false
      for (line in lines) {
        if (line.contains(ParseFinishStart)) {
          started = true
        } else if (line.contains(ParseFinishEnd)) {
          started = false
        } else if (started) {
          parseComplete = parseComplete + line + "\n"
        }
      }
      if (!started) {
        parseComplete = ParseFinishStart + "\n" + parseComplete + ParseFinishEnd
        content = content.replace(ParseFinishMark, parseComplete)
      }
    }
    content = "package " + packageName + "\n\n" + (defaultImports + "\n" + (listOf(imps, imports).flatten()).joinToString("\n")).split("\n").toSortedSet().joinToString("\n") + "\n\n\n" + content
    des.delete()
    des.createNewFile()
    val writer = PrintWriter(des)
    writer.write(content)
    writer.close()
  }

  fun writeFile(des: File, cont: String) {
    des.mkdirs()
    des.delete()
    des.createNewFile()
    val writer = PrintWriter(des)
    writer.write(cont)
    writer.close()
  }


  val os = ArrayList<ObjectType>()
  val ies = ArrayList<IntEnumType>()
  val ses = ArrayList<StringEnumType>()
  fun o(s: String, vararg fs: Field): ObjectType {
    val o = ObjectType(s, fs.toList())
    os.add(o)
    return o
  }

  operator fun String.div(str: String)  = StringEnumItem(this, str)

  operator fun String.div(str: Int) = IntEnumItem(this, str)

  fun stringEnum(s: String, de: Int, vararg fs: StringEnumItem): StringEnumType {
    val o = StringEnumType(s, fs.toList(), de)
    ses.add(o)
    return o
  }

  fun intEnum(s: String, de: Int, vararg fs: IntEnumItem): IntEnumType {
    val o = IntEnumType(s, fs.toList(), de)
    ies.add(o)
    return o
  }

  fun codegen() {
    println("codegen...")
    ies.forEach{ e -> writeClass(e.name, e.codgen())}
    ses.forEach{ e -> writeClass(e.name, e.codgen())}
    os.forEach{o -> writeClass(o.name, o.codgen())}
  }

}


object BaseSpec : Spec("org.snailya.demo.data", File("generator/src/main/java"), emptyList()) {

  @JvmStatic
  fun main(args: Array<String>): Unit = codegen()

  val stringToUri = ConvertedType(string, JvmType("Uri", "defaultUri").n, "stringToUri")

  init {

    stringEnum("SomeStringEnum",
        0,
        "CLASSIC" / "what",
        "LITE" / "jiji",
        "SPOCK" / "this"
    )
    intEnum("SomeEnum",
        0,
        "CLASSIC" / 1,
        "LITE" / 2,
        "SPOCK" / 3
    )

    o("Location",
        f("latitude", double).apply { isOverride = true },
        f("longitude", double),
        f("coordinateType", HandmadeObjectType("CoordinateType").n),
        f("countryCode", string.n),
        f("countryName", string.n),
        f("cityName", string.n),
        f("someArray", list(string)),
        f("someUrl", stringToUri),
        f("cached", boolean)
    ).apply {
      implements = "What"
    }
  }
}



